packages/promotion/src/services/promotion-module.ts

```diff
@@ -90,6 +91,120 @@ export default class PromotionModuleService<
     return joinerConfig
   }
 
+  @InjectManager("baseRepository_")
+  async registerUsage(
+    computedActions: PromotionTypes.ComputeActions[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<void> {
+    const promotionCodes = computedActions
+      .map((computedAction) => computedAction.code)
+      .filter(Boolean)
+
+    const promotionCodeCampaignBudgetMap = new Map<
+      string,
+      UpdateCampaignBudgetDTO
+    >()
+    const promotionCodeUsageMap = new Map<string, boolean>()
+
+    const existingPromotions = await this.list(
+      { code: promotionCodes },
+      { relations: ["application_method", "campaign", "campaign.budget"] },
+      sharedContext
+    )
+
+    const existingPromotionsMap = new Map<string, PromotionTypes.PromotionDTO>(
+      existingPromotions.map((promotion) => [promotion.code!, promotion])
+    )
+
+    for (let computedAction of computedActions) {
+      if (!ComputeActionUtils.canRegisterUsage(computedAction)) {
+        continue
+      }
+
+      computedAction = computedAction as PromotionTypes.UsageComputedActions
+      const promotion = existingPromotionsMap.get(computedAction.code)
+
+      if (!promotion) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          `Promotion with code ${computedAction.code} not found`
+        )
+      }
```

Comment:
thought: don't think we should fail here - e.g., if the code was deleted we don't want that to stand in the way of registering usage across all the other campaigns.

Reply:
If this is done as a part of the cart completion transaction, then the promo would have already been applied. Do we ignore that in this case?

Reply:
I would be more permissive at the module level and then more restrictive in the workflow - this way if you can use the module for both cases and configure it in the workflow